turbopack/crates/turbo-tasks/src/registry.rs

```diff
@@ -1,145 +1,161 @@
-        map_by_name.insert(global_name, new_id);
-        e.insert(new_id);
-        Some(new_id)
-    } else {
-        None
-    }
-}
-
-fn get_thing_id<K, V>(value: V, map_by_value: &FxDashMap<V, K>) -> K
-where
-    V: Hash + Eq + Debug,
-    K: Clone,
-{
-    if let Some(id) = map_by_value.get(&value) {
-        id.clone()
-    } else {
-        panic!("Use of unregistered {value:?}");
+pub fn get_function_by_global_name(global_name: &str) -> &'static NativeFunction {
+    static NAME_TO_FUNCTION: Lazy<FxHashMap<&'static str, &'static NativeFunction>> =
+        Lazy::new(|| {
+            let mut map = FxHashMap::default();
+            for collected in inventory::iter::<CollectableFunction> {
+                let native_function = &**collected.0;
+                let global_name = native_function.global_name;
+                let prev = map.insert(global_name, native_function);
+                debug_assert!(
+                    prev.is_none(),
+                    "multiple functions registered with the name {global_name}!"
+                );
+            }
+            map.shrink_to_fit();
+            map
+        });
+
+    match NAME_TO_FUNCTION.get(global_name) {
+        Some(f) => f,
+        None => panic!("unable to find function: {global_name}"),
     }
 }
 
-/// Registers a function so it is available for persistence
-pub fn register_function(global_name: &'static str, func: &'static NativeFunction) {
-    let prev = NAME_TO_FUNCTION.write().unwrap().insert(global_name, func);
-    debug_assert!(
-        prev.is_none(),
-        "registration mappings for {global_name} are inconsistent!"
-    );
+struct Values {
+    id_to_value: Box<[&'static ValueType]>,
+    value_to_id: FxHashMap<&'static ValueType, ValueTypeId>,
+    global_name_to_value: FxHashMap<&'static str, (ValueTypeId, &'static ValueType)>,
 }
 
-pub fn get_function_by_global_name(global_name: &str) -> &'static NativeFunction {
-    NAME_TO_FUNCTION.read().unwrap().get(global_name).unwrap()
-}
+static VALUES: Lazy<Values> = Lazy::new(|| {
+    // Inventory does not guarantee an order. So we sort by the global name to get a stable order
+    // This ensures that assigned ids are also stable.
+    // We don't currently take advantage of this but we could in the future.  The remaining issue is
+    // ensuring the set of values is the same across runs.
+    let mut all_values = inventory::iter::<CollectableValueType>
+        .into_iter()
+        .map(|t| &**t.0)
+        .collect::<Vec<_>>();
+    all_values.sort_by_key(|t| t.global_name);
+
+    let mut value_to_id = FxHashMap::default();
+    value_to_id.reserve(all_values.len());
+    let mut global_name_to_value = FxHashMap::default();
+    global_name_to_value.reserve(all_values.len());
+
+    for (index, &value_type) in all_values.iter().enumerate() {
+        // SAFETY: as a usize, usize+1 is definitely non-zero.
+        let id = unsafe { ValueTypeId::new_unchecked((index + 1).try_into().unwrap()) };
```

Comment:
Nit: Just from the standpoint of reducing type conversions, it's easier to keep your own `NonZeroU32` counter.

```suggestion

    let mut next_id = NonZeroU32::MIN;
    for &value_type in all_values.iter() {
        let id = ValueTypeId::from(id);
        next_id.checked_add(1).unwrap();
```

I'm guessing you're avoiding `IdFactory` because you don't need the atomic, but I don't think it would matter much, as an `AtomicU64` doesn't really cost anything when it's only used from a single thread.

Reply:
Nothing really against IdFactory it just seemed overkill once we switched to loops.

Good idea on switching to a direct counter